File: server/ncurses/widget.php
<?php

class Widget {
    protected $x;
    protected $y;
    private $width;
    private $height;
    private $parent;
    private $widgets = array();

    function __construct($x, $y, $width, $height, $parent) {
        $this->x = $x;
        $this->y = $y;
        $this->width = $width;
        $this->height = $height;

        // if ($parent != null) {
        //     $this->parent = $parent;
        //     $parent->add_child($this);
        // }
    }

    function __destruct() {
    }

    function add($widget) {
        array_push($this->widgets, $widget);
    }

    function setX($x) {
        $this->x = $x;
    }

    function setY($y) {
        $this->y = $y;
    }

    function draw($screen) {
        // clean area of the widget
        for ($y = 0; $y < $this->height; $y++) {
            $screen->move($this->x, $this->y + $y);
            $screen->printw(str_repeat(' ', $this->width));
        }
        // draw children
        foreach ($this->widgets as $widget) {
            $widget->draw($screen);
        }
    }
}

?>

File: server/ncurses/widget.py
class Widget:
    x = 0
    y = 0
    width = 0
    height = 0
    # parent
    widgets = []

    def __init__(self, x, y, width, height, parent):
        self.x = x
        self.y = y
        self.width = width
        self.height = height

        # if parent != None:
        #     self.parent = parent
        #     parent.add(self)

    def add(self, widget):
        self.widgets.append(widget)

    def setX(self, x):
        self.x = x

    def setY(self, y):
        self.y = y

    def draw(self, screen):
        # clean area of the widget
        for y in range(0, self.height):
            screen.move(self.x, self.y + y)
            screen.printw(' ' * self.width)

        # draw children
        for widget in self.widgets:
            widget.draw(screen)
